src/utils/tx_tools.py

```diff
@@ -0,0 +1,364 @@
+        for input in resolved_reference_inputs + resolved_inputs:
+            if input.output.script is not None:
+                potential_scripts.append(input.output.script)
+        try:
+            spending_script = next(
+                s
+                for s in tx.transaction_witness_set.plutus_v2_script
+                if plutus_script_hash(PlutusV2Script(s))
+                == spending_input.output.address.payment_part
+            )
+        except (StopIteration, TypeError):
+            raise NotImplementedError(
+                "Can not validate spending of non plutus v2 script (or plutus v2 script is not in context)"
+            )
+        if spending_input.output.datum is not None:
+            datum = spending_input.output.datum
+        elif spending_input.output.datum_hash is not None:
+            datum_h = spending_input.output.datum_hash
+            try:
+                datum = next(
+                    d
+                    for d in tx.transaction_witness_set.plutus_data or []
+                    if datum_hash(d) == datum_h
+                )
+            except StopIteration:
+                raise ValueError(
+                    f"No datum with hash '{datum_h.payload.hex()}' provided for transaction"
+                )
+        else:
+            raise ValueError(
+                "Spending input is missing an attached datum and can not be spent"
+            )
+        script_contexts.append(
+            ScriptInvocation(
+                spending_script,
+                datum,
+                spending_redeemer,
+                ScriptContext(tx_info, Spending(to_tx_out_ref(spending_input.input))),
+            )
+        )
+    for i, minting_script_hash in enumerate(tx.transaction_body.mint or []):
+        try:
+            minting_redeemer = next(
+                r
+                for r in tx.transaction_witness_set.redeemer
+                if r.index == i and r.tag == RedeemerTag.MINT
+            )
+        except StopIteration:
+            raise ValueError(
+                f"Missing redeemer for mint {i} (index or tag set incorrectly or missing redeemer)"
+            )
+        try:
+            minting_script = next(
+                s
+                for s in tx.transaction_witness_set.plutus_v2_script
+                if plutus_script_hash(PlutusV2Script(s)) == minting_script_hash
+            )
+        except StopIteration:
+            raise NotImplementedError(
+                "Can not validate spending of non plutus v2 script (or plutus v2 script is not in context)"
+            )
+
+        script_contexts.append(
+            ScriptInvocation(
+                minting_script,
+                minting_redeemer,
+                ScriptContext(tx_info, Minting(minting_script.hash())),
+            )
+        )
+    return script_contexts
+
+
+def evaluate_script(script_invocation: ScriptInvocation):
+    uplc_program = pyaiken.uplc.unflat(script_invocation.script_type.hex())
```

Comment:
You can play around here with a) caching the unflattened result and b) unflattening the script using the python native tools